### PR TITLE
Talon side spoken form tests

### LINKED ISSUE #449: Add regression tests for voice commands
We'd like to ensure that voice commands don't break.  Probably the most comprehensive way to do this would be to convert all of our cursorless VSCode [recorded tests](https://github.com/pokey/cursorless-vscode/tree/main/src/test/suite/fixtures/recorded) into voice command tests.  We'd basically take the `spokenForm` field of the test, call `mimic` on it, and ensure that its output matches the `command` in the test.  We'd need to do some bulk processing to fix the spoken forms that didn't come from knausj

This effort would require https://github.com/talonvoice/talon/issues/375 in order to run in CI, but will start by getting some tests running locally to function as a starting point for a conversation around our needs for a talon CI runner

See also [knausj test setup](https://github.com/knausj85/knausj_talon/blob/main/tests), esp:
- https://github.com/knausj85/knausj_talon/blob/main/tests/stubs/talon/__init__.py
- https://github.com/knausj85/knausj_talon/blob/main/tests/test_code_modified_function.py

Also maybe worth having a look at contract / hypothesis testing?
- https://github.com/mristin/icontract-hypothesis
- https://hypothesis.readthedocs.io/en/latest/index.html
- https://crosshair.readthedocs.io/en/latest/contracts.html

## Approach

- [x] Add Cursorless extension VSCode command `cursorless.generate_spoken_form_tests` that iterates recorded test cases and populates a tmp directory with spoken form tests, returning the path to that directory.  Might look at https://github.com/cursorless-dev/cursorless/blob/037389f87b16842e32511dbcffeca0d6e95969d2/packages/cursorless-engine/src/core/Cheatsheet.ts#L58 for inspiration
- [x] Add Talon tag `user.cursorless_default_vocabulary` that switches all spoken forms to their normal forms (eg stock knausj alphabet, no modifier overrides)
- [x] Add Talon action that does the following:
  1. Calls the `cursorless.generate_spoken_form_tests` VSCode action
  2. Enables `user.cursorless_default_vocabulary`
  3. Enables a tag that overrides the cursorless command actions from https://github.com/cursorless-dev/cursorless/blob/main/cursorless-talon/src/command.py to just assert that the payload about to be sent matches expectation, and then doesn't send the action
  4. Iterates through the generated tests and runs them using `mimic` on the spoken form, setting up the payload checker with the payload.  Possibly we should run 3) once per test case
  5. Disable both tags above
- [/] Possibly add spoken form for above action to `cursorless-talon-developer`
- [x] For tests that fail due to our own custom spoken forms, we will normalize them in the PR so that they're correct in the actual recorded test.  We can do that by using a [custom transformation](https://www.cursorless.org/docs/contributing/test-case-recorder/#custom-transformation) that just keeps a dictionary that maps from our spoken forms to default and keep adding to the dict as tests fail.  Could call it `normalizeSpokenForms` and check it in to version control so we can use it if necessary in the future if spoken forms change for some reason
- [/] Make it so that `"cursorless record"` enables the `user.cursorless_default_vocabulary` tag so that we don't end up recording tests with bad spoken forms in the future
- [/] Add checkbox to PR template saying "I have run spoken form tests"

### LINKED ISSUE #1675: Change "take vest past after air" test to "take vest past end of air"
Interesting. We should prob change this test to use "end of". Or just drop it 🤷‍♂️

_Originally posted by @pokey in https://github.com/cursorless-dev/cursorless/pull/1674#discussion_r1271428440_

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
- Fixes #449
- Fixes #1675

## Potential manual tests to add

For these, just bake payload and spoken form into `const`s in the test itself

- [x] "slice past"
- [x] "two tokens forward"
- [x] "one token"
- [x] "take inside pair"

## Checklist

- [x] Handle if Talon was in sleep mode initially
- [x] Add deprecation warning if user says "to after"?  I believe we support this today, and some users are still saying that, as indicated by slack questions from the past month.  We could probably have this as an alternate spoken form for "after" and show a warning if they use it
- [x] Stop using `generateSpokenForms` to check whether to run a test; instead read `spokenFormError` field on recorded test
- [x] How do we actually want to launch this? Launch config in vscode and then?
- [x] Add notifications when test start and stop
- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [/] I have updated the [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs)
- [x] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/core/commandVersionUpgrades/canonicalizeSpokenFormTestCommand.ts
Comment (pokey) [inline packages/cursorless-engine/src/core/commandVersionUpgrades/canonicalizeSpokenFormTestCommand.ts:1]: This is epic 😍

File: packages/common/src/definitions/modifiers.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+
+import { objectFromKeys, swapKeyAndValue } from "../util/objectUtils";
+
+const modifiers = {
+  bounds: "excludeInterior",
```
Comment (pokey) [inline packages/common/src/definitions/modifiers.ts:6]: I would have mapped id to spoken form, as the former is the primary key for that entity. How come you went this way?

**Review by pokey — COMMENTED** [review]
Just left a couple very quick comments

File: packages/common/src/definitions/modifiers.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/common/src/definitions/modifiers.ts:6]: This is the way we are used to seeing them in Talon. I'm also thinking that we're going to use the same source to actually generate the spoken forms as well as cheatsheet. In my mind this just feels "right".

File: packages/cursorless-engine/src/core/commandVersionUpgrades/canonicalizeSpokenFormTestCommand.ts
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/core/commandVersionUpgrades/canonicalizeSpokenFormTestCommand.ts:1]: Glad you like it :)

File: packages/common/src/definitions/modifiers.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/common/src/definitions/modifiers.ts:6]: All true.

**Comment by pokey** [comment]
Worth running through the checkboxes in these two issues to see if you've addressed everything.  If not, I'm not saying you should do it in this PR, but we should prob not close those issues:

- https://github.com/cursorless-dev/cursorless/issues/449
- https://github.com/cursorless-dev/cursorless/issues/1440

**Comment by AndreasArvidsson** [comment]
@pokey Updated and ready for review

**Comment by AndreasArvidsson** [comment]
> Ok this looks good to me, but for some reason is still failing for me locally, so we'll need to debug. Also worth having a look at the changes I made

The changes just looks fine. I just discover that after Talon has restarted some of the tests are failing for me as well. Touching a single Talon user file fixes this for some reason.

**Comment by AndreasArvidsson** [comment]
Its the Cursorless simple actions list that don't get updated to default unless you touch a file. I will investigate further

**Review by pokey — APPROVED** [review]

**Pull request merged**